corehq/apps/change_feed/management/commands/reconcile_producer_logs.py

```diff
@@ -0,0 +1,236 @@
+
+import settings
+from corehq.apps.change_feed.producer import (
+    CHANGE_ERROR,
+    CHANGE_PRE_SEND,
+    CHANGE_SENT,
+    KAFKA_AUDIT_LOGGER,
+)
+
+logger = logging.getLogger(KAFKA_AUDIT_LOGGER)
+
+
+class DocTypeReconciliation(object):
+    """Reconcile Kafka producer transactions. Each attempt to send has a unique
+    transaction ID which is used by the reconciliation.
+
+    Transactions are processed until point A at which stage the reconcilliation is 'frozen'.
+    Transactions seen after the 'freeze' will only be used to reconcile transactions seen before the freeze
+    and will not be added to the list of transactions to reconcile.
+    """
+    def __init__(self, doc_type):
+        self.doc_type = doc_type
+        self.pre_send = {}
+        self.error_doc_ids = set()
+        self.count = 0
+
+        self.frozen = False
+
+    def freeze(self):
+        self.frozen = True
+
+    def add_row(self, row):
+        date, trans_type, doc_type, doc_id, transaction_id = row
+        if doc_type != self.doc_type:
+            return
+
+        if trans_type == CHANGE_PRE_SEND and not self.frozen:
+            self.count += 1
+            self.pre_send[transaction_id] = doc_id
+        elif trans_type == CHANGE_SENT:
+            try:
+                del self.pre_send[transaction_id]
+            except KeyError:
+                pass
+            try:
+                self.error_doc_ids.remove(doc_id)
+            except KeyError:
+                pass
+        if trans_type == CHANGE_ERROR and (not self.frozen or transaction_id in self.pre_send):
+            try:
+                del self.pre_send[transaction_id]
+            except KeyError:
+                pass
+            self.error_doc_ids.add(doc_id)
+
+    def get_results(self):
+        return {
+            'transaction_count': self.count,
+            'persistent_error_count': len(self.error_doc_ids),
+            'unaccounted_for': len(self.pre_send),
+            'unaccounted_for_ids': set(self.pre_send.values()),
+        }
+
+    def has_results(self):
+        return self.error_doc_ids or self.pre_send
+
+
+class Reconciliation(object):
+    def __init__(self):
+        self.by_doc_type = {}
+
+    def freeze(self):
+        for doc_type_recon in self.by_doc_type.values():
+            doc_type_recon.freeze()
+
+    def add_row(self, row):
+        doc_type = row[2]
+        if doc_type not in self.by_doc_type:
+            self.by_doc_type[doc_type] = DocTypeReconciliation(doc_type)
+        self.by_doc_type[doc_type].add_row(row)
+
+    def reconcile(self):
+        for doc_type_recon in self.by_doc_type.values():
+            doc_type_recon.reconcile()
+
+    def get_results(self):
+        return {
+            doc_type: recon.get_results()
+            for doc_type, recon in self.by_doc_type.items()
+            if recon.has_results()
+        }
+
+
+class Command(BaseCommand):
+    help = """"Kafka Producer log file reconciliation command.
```

Comment:
This management command feels pretty complex to me and took a while for me to understand. From what I can tell it could be written as something similar to :

```bash
grep 'PRE-SEND' kafka-audit.log* | awk -F, '{print $4}' | sort > intended_transaction_ids.log
grep 'SENT' kafka-audit.log* | awk -F, '{print $4}' | sort > sent_transaction_ids.log
comm -23 intended_transaction_ids.log sent_transaction_ids.log > unsent_transaction_ids.log

grep 'ERROR' kafka-audit.log* | awk -F, '{print $4}' > errored_ids.log

grep -f unsent_transaction_ids.log kafka-audit.log* | awk -F, '{print $3}' > unsent_doc_ids.log
```

"transactions proccessed" would be  `wc -l sent_transaction_ids.log`
"Persistent errors" would be slightly changed as it would just be errors in general: `wc -l error_ids.log`
"unaccounted for transaction count" would be `wc -l unsent_transaction_ids.log`
"doc_ids unaccounted for" would also be slightly changed as it would be doc_ids that had at least one failed process. `unsent_doc_ids.log`

Does the above definition of things look correct? To me that's a bit faster to understand than the python code and I have to track fewer variables/functions. Though I'm not sure that we have a good system of when to implement something in bash vs python

Reply:
yes, something like that would work though it would be a bit more complex since if you just run it on all log files you end up with some false positives. It would require writing a bash script that I don't expect would be much less complex in the end and likely less well tested.

Reply:
It wouldn't necessarily require a bash script. We already have http://amoffat.github.io/sh/ in our requirements for some hyrid scripts we run. But anyways if its working now, there's not a good reason to re-write it.